clang/lib/CIR/Dialect/Transforms/RegionInlining.cpp

```diff
@@ -0,0 +1,308 @@
+                                                     body, exit);
+  }
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::LoopOpInterface op,
+                  mlir::PatternRewriter &rewriter) const override {
+    // Setup CFG blocks.
+    auto *entry = rewriter.getInsertionBlock();
+    auto *exit = rewriter.splitBlock(entry, rewriter.getInsertionPoint());
+    auto *cond = &op.getCond().front();
+    auto *body = &op.getBody().front();
+    auto *step = (op.maybeGetStep() ? &op.maybeGetStep()->front() : nullptr);
+
+    // Setup loop entry branch.
+    rewriter.setInsertionPointToEnd(entry);
+    rewriter.create<mlir::cir::BrOp>(op.getLoc(), &op.getEntry().front());
+
+    // Branch from condition region to body or exit.
+    auto conditionOp = cast<mlir::cir::ConditionOp>(cond->getTerminator());
+    lowerConditionOp(conditionOp, body, exit, rewriter);
+
+    // TODO(cir): Remove the walks below. It visits operations unnecessarily,
+    // however, to solve this we would likely need a custom DialecConversion
+    // driver to customize the order that operations are visited.
+
+    // Lower continue statements.
+    mlir::Block *dest = (step ? step : cond);
+    op.walkBodySkippingNestedLoops([&](mlir::Operation *op) {
+      if (isa<mlir::cir::ContinueOp>(op))
+        lowerTerminator(op, dest, rewriter);
+    });
+
+    // Lower break statements.
+    walkRegionSkipping<mlir::cir::LoopOpInterface, mlir::cir::SwitchOp>(
+        op.getBody(), [&](mlir::Operation *op) {
+          if (isa<mlir::cir::BreakOp>(op))
+            lowerTerminator(op, exit, rewriter);
+        });
+
+    // Lower optional body region yield.
+    // auto bodyYield = dyn_cast<mlir::cir::YieldOp>(body->getTerminator());
+    // if (bodyYield)
+    //   lowerTerminator(bodyYield, (step ? step : cond), rewriter);
+
+    for (auto& blk : op.getBody().getBlocks()) {
+      auto bodyYield = dyn_cast<mlir::cir::YieldOp>(blk.getTerminator());
+      if (bodyYield)
+        lowerTerminator(bodyYield, (step ? step : cond), rewriter);
+    }
+
+    // Lower mandatory step region yield.
+    if (step)
+      lowerTerminator(cast<mlir::cir::YieldOp>(step->getTerminator()), cond,
+                      rewriter);
+
+    // Move region contents out of the loop op.
+    rewriter.inlineRegionBefore(op.getCond(), exit);
+    rewriter.inlineRegionBefore(op.getBody(), exit);
+    if (step)
+      rewriter.inlineRegionBefore(*op.maybeGetStep(), exit);
+
+    rewriter.eraseOp(op);
+    return mlir::success();
+  }
+};
+
+
+
+void populateRegionInliningPatterns(RewritePatternSet &patterns) { 
+  // clang-format off
+  patterns.add<
+    IfOpRegionsInlining,
+    ScopeOpRegionsInlining,
+    LoopOpRegionsInlining
+  >(patterns.getContext());
+  // clang-format on
+}
+
+// TODO: this is still a draft, not sure it's a good way. Maybe create a separate
+// pass for it
+void RegionInliningPass::process_gotos(SmallVector<GotoOp>& gotos) {
```

Comment:
here is place where `GoTo` is replaced by CIR `BrOp`. Probably it's not a good place - and I can create a separate pass for that after `RegionInlining`